v3/tasks/20241204-v3-vault/index.ts

```diff
@@ -27,33 +27,53 @@ export default async (task: Task, { force, from }: TaskRunOptions = {}): Promise
 
   // Skip deployment if it's already done
   if ((await vaultFactory.isDeployed(vaultAddress)) === false) {
-    await vaultFactory.create(
-      input.salt,
-      vaultAddress,
-      input.vaultCreationCode,
-      input.vaultExtensionCreationCode,
-      input.vaultAdminCreationCode,
-      { gasLimit: 17e6 }
+    const deployTransaction = await task.deployFactoryContracts(
+      await vaultFactory.populateTransaction.create(
+        input.salt,
+        vaultAddress,
+        input.vaultCreationCode,
+        input.vaultExtensionCreationCode,
+        input.vaultAdminCreationCode,
+        { gasLimit: 17e6 }
+      ),
+      ['Vault', 'VaultExtension', 'VaultAdmin', 'ProtocolFeeController'],
+      from,
+      force
     );
-  }
 
-  const protocolFeeControllerAddress = await vaultFactory.deployedProtocolFeeControllers(vaultAddress);
-  const vaultExtensionAddress = await vaultFactory.deployedVaultExtensions(vaultAddress);
-  const vaultAdminAddress = await vaultFactory.deployedVaultAdmins(vaultAddress);
+    if (deployTransaction) {
+      const protocolFeeControllerAddress = await vaultFactory.deployedProtocolFeeControllers(vaultAddress);
+      const vaultExtensionAddress = await vaultFactory.deployedVaultExtensions(vaultAddress);
+      const vaultAdminAddress = await vaultFactory.deployedVaultAdmins(vaultAddress);
 
-  await task.verify('Vault', vaultAddress, [vaultExtensionAddress, input.Authorizer, protocolFeeControllerAddress]);
-  await task.verify('VaultExtension', vaultExtensionAddress, [vaultAddress, vaultAdminAddress]);
-  await task.verify('VaultAdmin', vaultAdminAddress, [
-    vaultAddress,
-    input.pauseWindowDuration,
-    input.bufferPeriodDuration,
-    input.minTradeAmount,
-    input.minWrapAmount,
-  ]);
-  await task.verify('ProtocolFeeController', protocolFeeControllerAddress, [vaultAddress]);
-
-  await task.save({ Vault: vaultAddress });
-  await task.save({ VaultExtension: vaultExtensionAddress });
-  await task.save({ VaultAdmin: vaultAdminAddress });
-  await task.save({ ProtocolFeeController: protocolFeeControllerAddress });
+      await task.saveAndVerifyFactoryContracts(deployTransaction, [
```

Comment:
There's one thing to consider here.

Usually we need to run the scripts more than once to effectively verify the contract.

The first time they are deployed, but the block explorer is not fast enough to verify the contracts right away. So what we usually do is just run the script again.

It's fine to call `task.save` more than once as it doesn't change anything, and if the contracts are already deployed we should just skip that part (that's what the basic `deployAndVerify` does). But we should execute the `verify` bit every time, not only the first one.